extensions/amp-social-share/0.2/social-share.js

```diff
@@ -30,63 +32,196 @@ const DEFAULT_HEIGHT = 44;
     }
+
+    // Verify width and height are valid integers
+    let checkedWidth = width || DEFAULT_WIDTH;
+    let checkedHeight = height || DEFAULT_HEIGHT;
+    if (width && !Number.isInteger(width)) {
+      throwWarning_(
+        `The width property should be an Integer, defaulting to ${DEFAULT_WIDTH}. ${name}`
+      );
+      checkedWidth = DEFAULT_WIDTH;
+    }
+    if (height && !Number.isInteger(height)) {
+      throwWarning_(
+        `The height property should be an Integer, defaulting to ${DEFAULT_HEIGHT}. ${name}`
+      );
+      checkedHeight = DEFAULT_HEIGHT;
+    }
+    return {
+      typeConfig,
+      baseEndpoint,
+      checkedWidth,
+      checkedHeight,
+      obsoleteType,
+    };
   }
 
-  /** @private */
-  function handleActivation() {
-    const href = props['href'];
-    const target = props['target'] || '_blank';
-    if (!href) {
-      throw new Error('Clicked before href is set.');
+  /**
+   * @private
+   * @param {!JsonObject} typeConfig
+   * @param {?string} baseEndpoint
+   * @param {!JsonObject} props
+   * @return {?string}
+   */
+  function createEndpoint_(typeConfig, baseEndpoint, props) {
+    const {params, bindings} = props;
+    const combinedParams = dict();
+    Object.assign(combinedParams, typeConfig['defaultParams'], params);
+    const endpointWithParams = addParamsToUrl(baseEndpoint, combinedParams);
+
+    const combinedBindings = dict();
+    const bindingVars = typeConfig['bindings'];
+    if (bindingVars) {
+      bindingVars.forEach((name) => {
+        combinedBindings[name.toUpperCase()] = params[name] ? params[name] : '';
+      });
     }
-    if (startsWith(href, 'navigator-share:')) {
-      if (!navigator.share) {
-        throw new Error('no navigator.share');
-      }
-      const dataStr = href.substr(href.indexOf('?'));
-      const data = parseQueryString(dataStr);
-      navigator.share(data).catch(() => {
-        // TODO(alanorozco): Warn here somehow.
-        // warn(TAG, e.message, dataStr);
+    if (bindings) {
+      Object.keys(bindings).forEach((name) => {
+        combinedBindings[name.toUpperCase()] = bindings[name]
+          ? bindings[name]
+          : '';
       });
+    }
+    const finalEndpoint = Object.keys(combinedBindings).reduce(
+      (endpoint, binding) =>
+        endpoint.replace(new RegExp(binding, 'g'), combinedBindings[binding]),
+      endpointWithParams
+    );
+    return finalEndpoint;
+  }
+
+  /**
+   * @private
+   * @param {?string} message
+   */
+  function throwError_(message) {
+    throw new Error(message);
+  }
+
+  /**
+   * @private
+   * @param {?string} message
+   */
+  function throwWarning_(message) {
+    console.warn(message);
+  }
+
+  /**
+   * @private
+   */
+  function handleActivation_() {
+    if (finalEndpoint.split(':', 1)[0] === 'navigator-share') {
+      if (window && window.navigator && window.navigator.share) {
+        const dataStr = finalEndpoint.substr(finalEndpoint.indexOf('?'));
```

Comment:
There should be a util for this in the `url.js`.

Reply:
Looks like there is getFragment() which gets params after '#' symbol but nothing for '?' unless I'm missing somehting.

Reply:
Though I'm not strictly against doing URL splitting here directly. But then we need to handle edge-cases, such as what if there's also a `#` in this URL, what if there's no `?`, etc.

Reply:
Thanks for the calling this out.  I was only able to find parseUrlDeprecated which I am hesitant to use because of the Deprecation.  I have attempted to write my own helper function (getQueryString, line 193), which handles the corner cases that I can think of.

No ? -> returns '' (empty string)
? and # -> only returns the substring from ? to #
# then ? (reversed order) -> returns substring from ? to the end of string
? and no # -> returns substring from ? to the end of string